[SPARK-7727] [SQL] Avoid inner classes in RuleExecutor

stephankessler opened the pull request:
Moved (case) classes Strategy, Once, FixedPoint and Batch to the companion object. This is necessary if we want to have the Optimizer easily extendable in the following sense: Usually a user wants to add additional rules, and just take the ones that are already there. However, inner classes made that impossible since the code did not compile

This allows easy extension of existing Optimizers see the DefaultOptimizerExtendableSuite for a corresponding test case.

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #47296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47296/consoleFull)** for PR 10174 at commit [`fb4988a`](https://github.com/apache/spark/commit/fb4988ab16b04753c8e8272de7227a325d22fc4c).

SparkQA commented:
**[Test build #47296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47296/consoleFull)** for PR 10174 at commit [`fb4988a`](https://github.com/apache/spark/commit/fb4988ab16b04753c8e8272de7227a325d22fc4c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47296/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47336/
Test FAILed.

SparkQA commented:
**[Test build #47337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47337/consoleFull)** for PR 10174 at commit [`707ab3d`](https://github.com/apache/spark/commit/707ab3d30afcfc73f12ea049ff468d2b100fde52).

SparkQA commented:
**[Test build #47337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47337/consoleFull)** for PR 10174 at commit [`707ab3d`](https://github.com/apache/spark/commit/707ab3d30afcfc73f12ea049ff468d2b100fde52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47337/
Test PASSed.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.sql.catalyst.analysis
 
+import org.apache.spark.sql.catalyst.rules.RuleExecutor.{Batch, Once, FixedPoint}
```
import ordering

stephankessler replied:
Done!

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala:
```diff
@@ -56,12 +55,16 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
 
   /** A batch of rules. */
-  protected case class Batch(name: String, strategy: Strategy, rules: Rule[TreeType]*)
+  case class Batch[TreeType <: TreeNode[_]]
```
I'd probably just make these top level classes if we are going to pull them out.  There is not a lot of stuff in the `rules` package.

stephankessler replied:
Got it, i moved Batch and Strategy to the rules package, and made Once and FixedPoint classes of the companion object of Strategy.

SparkQA commented:
**[Test build #47425 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47425/consoleFull)** for PR 10174 at commit [`bc39e03`](https://github.com/apache/spark/commit/bc39e03d64d0ac338a12531e5cc4433acd18bdd0).

SparkQA commented:
**[Test build #47425 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47425/consoleFull)** for PR 10174 at commit [`bc39e03`](https://github.com/apache/spark/commit/bc39e03d64d0ac338a12531e5cc4433acd18bdd0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Batch[TreeType <: TreeNode[_]] (name: String, strategy: Strategy, rules: Rule[TreeType]*)`\n  * `  case class FixedPoint(maxIterations: Int) extends Strategy`\n  * `abstract class Strategy `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47425/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/Strategy.scala:
```diff
@@ -0,0 +1,33 @@
+
+object Strategy {
+  /** A strategy that only runs once. */
+  case object Once extends Strategy { val maxIterations = 1 }
```
why do we put these 2 `case object` inside `object Strategy`?

stephankessler replied:
First i thought about making both classes on the top level, however ```org.apache.spark.sql.catalyst.rules.Once``` is not descriptive, while ```org.apache.spark.sql.catalyst.rules.Strategy.Once``` is. That was my idea :)

marmbrus commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/DefaultOptimizerExtendableSuite.scala:
```diff
@@ -0,0 +1,45 @@
+    * This class represents a dummy extended optimizer that takes the rules of the
+    * DefaultOptimizer and adds custom ones.
+    */
+  class ExtendedOptimizer extends Optimizer{
```
Nit: space before `{`

stephankessler replied:
Done

cloud-fan replied:
These is still no space before `{`?

stephankessler replied:
In my latest commit, there is one, in order to avoid confusing i (rebased) and squashed the commits to a single one!

marmbrus replied:
Hmm, weird.  Github is usually smarter than this.

marmbrus replied:
BTW, we squash automatically when merging.

SparkQA commented:
**[Test build #47577 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47577/consoleFull)** for PR 10174 at commit [`99968bc`](https://github.com/apache/spark/commit/99968bc91c2e70209b55ff17c49979f528c457bf).

SparkQA commented:
**[Test build #47577 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47577/consoleFull)** for PR 10174 at commit [`99968bc`](https://github.com/apache/spark/commit/99968bc91c2e70209b55ff17c49979f528c457bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Batch[TreeType <: TreeNode[_]] (name: String, strategy: Strategy, rules: Rule[TreeType]*)`\n  * `  case class FixedPoint(maxIterations: Int) extends Strategy`\n  * `abstract class Strategy `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47577/
Test PASSed.

SparkQA commented:
**[Test build #47648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47648/consoleFull)** for PR 10174 at commit [`dfa642a`](https://github.com/apache/spark/commit/dfa642ae920ecf5f1a870397e15d282e001c05b7).

SparkQA commented:
**[Test build #47648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47648/consoleFull)** for PR 10174 at commit [`dfa642a`](https://github.com/apache/spark/commit/dfa642ae920ecf5f1a870397e15d282e001c05b7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Batch[TreeType <: TreeNode[_]] (name: String, strategy: Strategy, rules: Rule[TreeType]*)`\n  * `  case class FixedPoint(maxIterations: Int) extends Strategy`\n  * `abstract class Strategy `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47648/
Test PASSed.

SparkQA commented:
**[Test build #47721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47721/consoleFull)** for PR 10174 at commit [`fea0c58`](https://github.com/apache/spark/commit/fea0c58805f20eb4a44e58181e63e45e67bd5a23).

3ourroom commented:
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-7727] [SQL] Avoid inner classes in RuleExecutor (#10174)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

gatorsmile commented:
Just curious. Is this for your project presented in "The Pushdown of Everything"?

stephankessler commented:
Just related, currently preparing appropriate documents to propose that to the community! Keep you posted!

SparkQA commented:
**[Test build #47721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47721/consoleFull)** for PR 10174 at commit [`fea0c58`](https://github.com/apache/spark/commit/fea0c58805f20eb4a44e58181e63e45e67bd5a23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Batch[TreeType <: TreeNode[_]] (name: String, strategy: Strategy, rules: Rule[TreeType]*)`\n  * `  case class FixedPoint(maxIterations: Int) extends Strategy`\n  * `abstract class Strategy `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47721/
Test PASSed.

gatorsmile commented:
Thank you! Looking forward to reading your proposal! : )

marmbrus commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/DefaultOptimizerExtendableSuite.scala:
```diff
@@ -0,0 +1,45 @@
+/**
+  * This is a test for SPARK-7727 if the Default Optimizer is kept being extendable
+  */
+class DefaultOptimizerExtendableSuite extends SparkFunSuite {
```
If this is the goal, I think this change could be a lot smaller (and thus a lot less disruptive to people who are already extending catalyst).

```
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
index f608869..f52dab1 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
@@ -28,10 +28,8 @@ import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.rules._
 import org.apache.spark.sql.types._
 
-abstract class Optimizer extends RuleExecutor[LogicalPlan]
-
-object DefaultOptimizer extends Optimizer {
-  val batches =
+abstract class Optimizer extends RuleExecutor[LogicalPlan] {
+  def batches =
     // SubQueries are only needed for analysis and can be removed before execution.
     Batch("Remove SubQueries", FixedPoint(100),
       EliminateSubQueries) ::
@@ -68,6 +66,11 @@ object DefaultOptimizer extends Optimizer {
       ConvertToLocalRelation) :: Nil
 }
 
+object DefaultOptimizer extends Optimizer {
+  val myBatches = Nil
+  override def batches = super.batches ++ myBatches
+}
+
 /**
  * Pushes operations down into a Sample.
  */
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala
index f80d2a9..62ea731 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala
@@ -59,7 +59,7 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   protected case class Batch(name: String, strategy: Strategy, rules: Rule[TreeType]*)
 
   /** Defines a sequence of rule batches, to be overridden by the implementation. */
-  protected val batches: Seq[Batch]
+  protected def batches: Seq[Batch]
 
```

stephankessler replied:
That is true, i investigated it and proposed another change

stephankessler committed:
* SPARK-7727 Avoid inner classes in RuleExecutor

This change makes sure that the Optimizer (and the Batches) are
kept extendable. One can create a different optimizer with a modified
batch set.

Inner classes are kept to keep changes minimal

SparkQA commented:
**[Test build #48002 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48002/consoleFull)** for PR 10174 at commit [`8b1a0f4`](https://github.com/apache/spark/commit/8b1a0f4f4500c51f45ae90882ae9e32495384bc3).

SparkQA commented:
**[Test build #48004 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48004/consoleFull)** for PR 10174 at commit [`78072af`](https://github.com/apache/spark/commit/78072af9026e5e91702c73faa85822b03df5ca95).

SparkQA commented:
**[Test build #48002 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48002/consoleFull)** for PR 10174 at commit [`8b1a0f4`](https://github.com/apache/spark/commit/8b1a0f4f4500c51f45ae90882ae9e32495384bc3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  * Abstract class all optimizers should inherit of, contains the standard batches (extending`\n  * `abstract class Optimizer extends RuleExecutor[LogicalPlan] `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48002/
Test PASSed.

SparkQA commented:
**[Test build #48004 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48004/consoleFull)** for PR 10174 at commit [`78072af`](https://github.com/apache/spark/commit/78072af9026e5e91702c73faa85822b03df5ca95).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  * Abstract class all optimizers should inherit of, contains the standard batches (extending`\n  * `abstract class Optimizer extends RuleExecutor[LogicalPlan] `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48004/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -28,44 +28,53 @@ import org.apache.spark.sql.catalyst.plans.logical._
+      Batch("Decimal Optimizations", FixedPoint(100),
+        DecimalAggregates) ::
+      Batch("LocalRelation", FixedPoint(100),
+        ConvertToLocalRelation) :: Nil
```
can you undo these indent changes?

stephankessler replied:
Done!

SparkQA commented:
**[Test build #48105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48105/consoleFull)** for PR 10174 at commit [`a7c9f5c`](https://github.com/apache/spark/commit/a7c9f5cc10234d68c84a59eb969e0e4cf3d84cf0).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -28,11 +28,13 @@ import org.apache.spark.sql.catalyst.plans.logical._
-  val batches =
-    // SubQueries are only needed for analysis and can be removed before execution.
+/**
+  * Abstract class all optimizers should inherit of, contains the standard batches (extending
```
nit: add enclosing bracket

marmbrus replied:
closing parentheses?

cloud-fan replied:
yea, parentheses `()`

stephankessler replied:
Done!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -28,11 +28,13 @@ import org.apache.spark.sql.catalyst.plans.logical._
+  * Optimizers can override this.
+  */
+abstract class Optimizer extends RuleExecutor[LogicalPlan] {
+  def batches: Seq[Batch] =
```
we can still use `val` here

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -28,11 +28,13 @@ import org.apache.spark.sql.catalyst.plans.logical._
+  */
+abstract class Optimizer extends RuleExecutor[LogicalPlan] {
+  def batches: Seq[Batch] =
+  // SubQueries are only needed for analysis and can be removed before execution.
```
add 2 space before this comment, to align it to the "Remove SubQueries" batch it's talking about.

stephankessler replied:
Good point! Thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -69,6 +71,13 @@ object DefaultOptimizer extends Optimizer {
+  * Non-abstract representation of the standard Spark optimizing strategies
+  */
+object DefaultOptimizer extends Optimizer {
+  override def batches: Seq[Batch] = super.batches
```
Do we need this override?

marmbrus replied:
I agree, this looks like a no-op.

stephankessler replied:
I changed it and modified the comment why we inherit!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala:
```diff
@@ -59,7 +59,7 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
 
   /** Defines a sequence of rule batches, to be overridden by the implementation. */
-  protected val batches: Seq[Batch]
+  protected def batches: Seq[Batch]
```
we can still use `val` here

marmbrus replied:
I don't think thats true.  You can't call `super` on a val.

cloud-fan replied:
Did a quick test, that's true, please ignore my previous comment.

SparkQA commented:
**[Test build #48105 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48105/consoleFull)** for PR 10174 at commit [`a7c9f5c`](https://github.com/apache/spark/commit/a7c9f5cc10234d68c84a59eb969e0e4cf3d84cf0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48105/
Test PASSed.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerExtendableSuite.scala:
```diff
@@ -0,0 +1,54 @@
+
+    // rules set to DummyRule, would not be executed anyways
+    val myBatches: Seq[Batch] = Batch("once", Once, DummyRule) ::
+                                Batch("fixedPoint", FixedPoint(100), DummyRule) :: Nil
```
You can follow `Optimizer` for how to indent these `Batch`es

stephankessler replied:
Done!

stephankessler committed:
* style and indent corrections

SparkQA commented:
**[Test build #48175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48175/consoleFull)** for PR 10174 at commit [`c7ea3b1`](https://github.com/apache/spark/commit/c7ea3b115540c8b1d428a56bb30434461e11f3e5).

SparkQA commented:
**[Test build #48175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48175/consoleFull)** for PR 10174 at commit [`c7ea3b1`](https://github.com/apache/spark/commit/c7ea3b115540c8b1d428a56bb30434461e11f3e5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  * Abstract class all optimizers should inherit of, contains the standard batches (extending`\n  * `abstract class Optimizer extends RuleExecutor[LogicalPlan] `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48175/
Test FAILed.

SparkQA commented:
**[Test build #48200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48200/consoleFull)** for PR 10174 at commit [`8d36e38`](https://github.com/apache/spark/commit/8d36e386948defd17839323c697e62c2d1fb878c).

SparkQA commented:
**[Test build #48200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48200/consoleFull)** for PR 10174 at commit [`8d36e38`](https://github.com/apache/spark/commit/8d36e386948defd17839323c697e62c2d1fb878c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class AssertNotNull(`\n  * `  * Abstract class all optimizers should inherit of, contains the standard batches (extending`\n  * `abstract class Optimizer extends RuleExecutor[LogicalPlan] `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48200/
Test PASSed.

marmbrus commented:
Thanks, merging to master.
